File: 백준/Silver/20922. 겹치는 건 싫어/겹치는 건 싫어.java
import java.util.*;
import java.io.*;

public class Main {

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    
    StringTokenizer st = new StringTokenizer(br.readLine());
    int N = Integer.parseInt(st.nextToken());
    int K = Integer.parseInt(st.nextToken());

    int[]seq = new int[N];
    st = new StringTokenizer(br.readLine());

    for (int i = 0; i < N; i++) {
      seq[i] = Integer.parseInt(st.nextToken());
    }

    int left = 0;
    int right = 0;
    int maxValue = 0;
    Map<Integer, Integer> dict = new HashMap<Integer, Integer>();

    while (left <= right && right < N) {
      int rightIdxFreq = dict.getOrDefault(seq[right],0);
      int leftIdxFreq = dict.getOrDefault(seq[left],0);
      if (rightIdxFreq < K) {
        dict.put(seq[right], rightIdxFreq+1);
        right++;
      } else {
        dict.put(seq[left], leftIdxFreq-1);
        left++;
      }
      maxValue = Math.max(maxValue, right-left);
    }

    System.out.println(maxValue);
  }
  
}

File: 백준/Silver/20922. 겹치는 건 싫어/겹치는 건 싫어.py
from collections import defaultdict 

N, K = map(int,input().split())
seq = list(map(int,input().split()))

max_seq  = 0

# left, right 를 벌려가면서 조건 확인
left, right = 0, 0 
dict = defaultdict(int)

while left <= right and right < N:
  #조건을 만족하면 right 증가 
  if dict[seq[right]] < K:
      dict[seq[right]] += 1
      right += 1
  else:
     dict[seq[left]] -= 1
     left += 1
  max_seq = max(max_seq, right - left)    

print(max_seq)


